constraint/blueprint.go

```diff
@@ -18,11 +18,16 @@ type Blueprint interface {
 	// NbOutputs return the number of output wires this blueprint creates.
 	NbOutputs(inst Instruction) int
 
-	// WireWalker returns a function that walks the wires appearing in the blueprint.
+	// WireWalker returns a function that walks the wires appearing in the blueprint
+	// and optionally the minimum level (minus one) at which the instruction can be solved;
+	// that is; maxLevel = max{level(wire) | wire in blueprint}.
+	// If maxLevel is unknown at compile time, implementation must return 0.
 	// This is used by the level builder to build a dependency graph between instructions.
-	WireWalker(inst Instruction) func(cb func(wire uint32))
+	WireWalker(inst Instruction) (w WireWalker, maxLevel int)
 }
 
+type WireWalker func(cb func(wire uint32) int)
```

Comment:
I'm not understanding the return value of `cb` here - is it `cb`'s definition of wire level? Does it allow the blueprint to learn the level of the wire from the system's perspective?

Reply:
To confirm - essentially we use the return value of WireWalker and `maxLevel` to keep the blueprint and system in sync what is the max level of solved wires. And when we update the level in the blueprint side then we assume that the wires for which the callback is called on should not be called again, and we internally update the starting conditions of the loops.